src/preferences/preferences.hpp

```diff
@@ -0,0 +1,861 @@
+		 * @param max_progress The value when the achievement is considered completed.
+		 * @param amount The amount to progress the achievement.
+		 * @return The achievement's current progress, or -1 if it has already been completed.
+		 */
+		int progress_achievement(const std::string& content_for, const std::string& id, int limit = 999999, int max_progress = 999999, int amount = 0);
+
+		/**
+		 * @param content_for The achievement group the achievement is part of.
+		 * @param id The ID of the achievement within the achievement group.
+		 * @param sub_id The ID of the sub-achievement within the achievement.
+		 * @return True if the sub-achievement exists and is completed, false otherwise.
+		 */
+		bool sub_achievement(const std::string& content_for, const std::string& id, const std::string& sub_id);
+
+		/**
+		 * Marks the specified sub-achievement as completed.
+		 *
+		 * Note that this uses the same [in_progress] as is used for progress_achievement().
+		 *
+		 * @param content_for The achievement group the achievement is part of.
+		 * @param id The ID of the achievement within the achievement group.
+		 * @param sub_id The ID of the sub-achievement within the achievement.
+		 */
+		void set_sub_achievement(const std::string& content_for, const std::string& id, const std::string& sub_id);
+
+		/**
+		 * @param addon_id The chosen addon id from the editor to store in the preferences.
+		 */
+		void set_editor_chosen_addon(const std::string& addon_id);
+
+		/**
+		 * @return The most recently selected add-on id from the editor. May be an empty string.
+		 */
+		std::string editor_chosen_addon();
+
+		void set_mp_alert_option(const std::string& id, const std::string& type, bool value);
+		bool mp_alert_option(const std::string& id, const std::string& type, bool def = false);
+		bool has_mp_alert_option(const std::string& id, const std::string& type);
+
+		void set_last_cache_cleared_version(const std::string& version);
+		std::string last_cache_cleared_version();
+
+		bool get_show_deprecation(bool def);
+
+		bool get_scroll_when_mouse_outside(bool def);
+
+		void set_dir_bookmarks(const config& cfg);
+		optional_const_config dir_bookmarks();
+
+		bool whisper_friends_only();
+		void set_whisper_friends_only(bool v);
+
+		bool auto_open_whisper_windows();
+
+		bool fi_invert();
+		void set_fi_invert(bool value);
+
+		bool fi_vacant_slots();
+		void set_fi_vacant_slots(bool value);
+
+		bool fi_friends_in_game();
+		void set_fi_friends_in_game(bool value);
+
+		bool fi_blocked_in_game();
+		void set_fi_blocked_in_game(bool value);
+
+		int auto_update_transitions();
+		void set_auto_update_transitions(int value);
+
+		std::string default_dir();
+
+		bool draw_terrain_codes();
+		void set_draw_terrain_codes(bool value);
+
+		bool draw_hex_coordinates();
+		void set_draw_hex_coordinates(bool value);
+
+		bool draw_num_of_bitmaps();
+		void set_draw_num_of_bitmaps(bool value);
```

Comment:
I accept the point about having a separate `preferences::editor::` namespace being different to the other preferences, but putting them all into a flat list like this loses all hints that these are settings that are only used in the editor.

I also think it would be a bad idea to unify the editor's persistent-across-restarts `draw_terrain_codes()` with the in-game but temporary `:draw_terrain_codes` commands.

Reply:
There's a console command `:show_terrain_codes` (my mistake to call it `:draw_terrain_codes`), which is similar to what the editor does, but isn't saved between sessions. I meant it would be bad to have the editor's preference also turn it on or off during normal gameplay.

Reply:
I don't think this PR changes that though?